perf(allocator/vec2): replace `self.reserve(1)` calls with `self.grow_one()` for better efficiency

Dunqing opened the pull request:
The places calling `self.reserve(1)` already checked `len == self.buf.cap()`, in order to avoid checking again in `reserve`, we should call `grow_one` instead which is a shortcut of `grow_amortized(len, 1)`.

Dunqing commented:
Unbelievable 😢
<img width="938" alt="image" src="https://github.com/user-attachments/assets/3fdca453-e69f-4175-88f3-6ba84dccb77f" />

Dunqing commented:
I've noticed the current `extend` implementation will call the `push` method, `extend` is quite a common method for our scenario, so I think `#[inline(never)]` is a deoptimization. I decided to get #9881 done first and then continue this.

overlookmotel commented:
### Merge activity

* **May 3, 10:33 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
perf(allocator/vec2): replace `self.reserve(1)` calls with `self.grow_one()` for better efficiency (#9856)

The places calling `self.reserve(1)` already checked `len == self.buf.cap()`, in order to avoid checking again in `reserve`, we should call `grow_one` instead which is a shortcut of `grow_amortized(len, 1)`.
